scripts/plugins/extauth-hook-AD.py

```diff
@@ -233,6 +223,55 @@ def _install(self):
                 os.remove(self._file_path)
 
 
+class UsersList(DynamicPam):
+    #pylint: disable=too-few-public-methods
+    """Class manage users which permit pool admin ssh"""
+
+    def __init__(self, session, arg, ad_enabled=True):
+        super(UsersList, self).__init__(HCP_USERS, session, arg, ad_enabled)
+
+    def _match_subject(self, subject_rec):
+        return subject_rec["other_config"]["subject-is-group"] != "true"
+
+    def _add_subject(self, subject_rec):
+        try:
+            sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            formatted_name = self._format_item(name)
+            logger.debug("Permit user %s with sid %s", formatted_name, sid)
+            self._lines.append(formatted_name)
+            # If ssh key is permittd in authorized_keys,
+            # The original name is compared, add UPN and original name
+            self._lines.append(subject_rec["other_config"]["subject-upn"])
```

Comment:
Does the `half-written`  means `formatted_name` is written, but original name and UPN does not write?
That actually does not matter.   And ```upn=subject_rec["other_config"]["subject-upn"] self._lines.append(upn)``` does not help on this.  But it is a good point that in case of subject does not have `subject-upn`, we should put the original name first.